crates/l2/sequencer/proof_coordinator.rs

```diff
@@ -0,0 +1,355 @@
+                                    tokio::task::spawn(async move {
+                                        if let Err(e) = self_clone.handle_connection(stream).await {
+                                            error!("Error handling connection from {addr}: {e}");
+                                        } else {
+                                            debug!("Connection from {addr} handled successfully");
+                                        }
+                                        drop(permit);
+                                    });
+                                },
+                                Err(e) => {
+                                    match e {
+                                        TryAcquireError::Closed => error!("Fatal error the semaphore has been closed: {e}"),
+                                        TryAcquireError::NoPermits => warn!("Connection limit reached. Closing connection from {addr}."),
+                                    }
+                                }
+                            }
+
+                        }
+                        Err(e) => {
+                            error!("Failed to accept connection: {e}");
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_connection(&self, mut stream: TcpStream) -> Result<(), ProverServerError> {
+        let mut buffer = Vec::new();
+        stream.read_to_end(&mut buffer).await?;
+
+        let data: Result<ProofData, _> = serde_json::from_slice(&buffer);
+        match data {
+            Ok(ProofData::BlockRequest) => {
+                if let Err(e) = self.handle_request(&mut stream).await {
+                    error!("Failed to handle request: {e}");
+                }
+            }
+            Ok(ProofData::ProofSubmit {
+                block_number,
+                calldata,
+            }) => {
+                if let Err(e) = self
+                    .handle_submit(&mut stream, block_number, calldata)
+                    .await
+                {
+                    error!("Failed to handle submit: {e}");
+                }
+            }
+            Err(e) => {
+                warn!("Failed to parse request: {e}");
+            }
+            _ => {
+                warn!("Invalid request");
+            }
+        }
+
+        debug!("Connection closed");
+        Ok(())
+    }
+
+    async fn handle_request(&self, stream: &mut TcpStream) -> Result<(), ProverServerError> {
+        info!("Request received");
+
+        let block_to_verify = 1 + EthClient::get_last_verified_block(
+            &self.eth_client,
+            self.on_chain_proposer_address,
+        )
+        .await?;
+
+        let latest_block_number = self.store.get_latest_block_number().await?;
+
+        let response = if block_to_verify > latest_block_number {
+            let response = ProofData::block_response(None, None);
```

Comment:
Maybe we can add a builder method with a descriptive name to build  `ProofData::block_response(None, None)`, which I understand is the response for a request that couldn't be fulfilled.

Reply:
Done [2c4afe3](https://github.com/lambdaclass/ethrex/pull/2478/commits/2c4afe38a5c13e46202b140c84cc948f99b888ee)